packages/connect/src/protocol-connect/get-request.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Message, protoBase64 } from "@bufbuild/protobuf";
+import type { AnyMessage } from "@bufbuild/protobuf";
+import type { UnaryRequest } from "../index.js";
+import {
+  headerContentType,
+  headerProtocolVersion,
+  headerUnaryAcceptEncoding,
+  headerUnaryContentLength,
+  headerUnaryEncoding,
+} from "./headers.js";
+
+const contentTypePrefix = "application/";
+
+function encodeMessageForUrl(message: Uint8Array, useBase64: boolean): string {
+  if (useBase64) {
+    // TODO(jchadwick-buf): Three regex replaces seems excessive.
+    // Can we make protoBase64.enc more flexible?
+    return protoBase64
+      .enc(message)
+      .replace(/\+/g, "-")
+      .replace(/\//g, "_")
+      .replace(/=+$/, "");
+  } else {
+    return encodeURIComponent(new TextDecoder().decode(message));
```

Comment:
I am afraid this is lossy. TextDecoder uses UTF-8 by default and will use replacement characters for unrecognized code points.

Reply:
It's probably not great that it isn't obvious the base64 path is ALWAYS taken for binary data, but since it is, I'm marking this resolved. If I can make the control flow clearer that would be nice, I'm open to suggestions.

Reply:
I suggest a doc comment on `transformConnectPostToGetRequest()`:

> If `message` is not valid UTF-8, `useBase64` must be true.

And to enforce it with `new TextDecoder("utf-8", {fatal: true})`.